### ISSUE TITLE
FEATURE - Consistent Script Execution Time

### ISSUE THREAD

**Issue by bauzer714** [boredazfcuk/docker-icloudpd#91]
The current implementation results in a sliding script start time as a result of synchronisation_interval being used to fill the time from "previous run end" and "next run start".

Example:
I want the script to run each day at 1 AM to enable dependencies to complete before the morning.

In the current implementation if the container started at 1am.

- 1am on Day 1
- 1:10 on Day 2
- 1:25 on Day 3
- ....

Before I started on a PR I wanted to see what your thoughts were.

- Supportive of this idea?
- Add an additional variable to enable this? -OR-
- Implement the behavior for all users

**Comment by boredazfcuk** [comment]
When I first implemented the script, I ran it on a crontab, it was just a little more complicated to configure hours and sync periods and such… so just I went for a loop.

I could re-implement from, but with a randomised minute (to prevent people hammering Apple on the hour, or whatever)

I could also implement a single_run variable so the script exits after the first run. That way, people could add a crontab on the host machine, like `docker exec -it --rm --env single_run=True <plus more variables> /usr/local/bin/sync-icloud.sh` which will create and destroy the container after each run.

Only problem then is the health check may have issues. Will have a think about how best to tackle it when I get home.

**Comment by bauzer714** [comment]
I had the same thoughts. A code based option could be:

1. Set start variable as epoch 
2. Run sync
3. Sleep time =  Interval - now epoch - start 
4. Execute sleep

**Comment by bauzer714** [comment]
My primary issue is the sliding start. Middle of the night is
 possible though with the startup delay.  There are a lot of use cases that are possible to implement. It needs to be a balance of use cases and what direction you are comfortable with the code base going.

**Comment by boredazfcuk** [comment]
I’ll implement whatever people want/need. Somebody did request cron based sync a while back, so I could look at something.

Only issue is making sure stuff doesn’t conflict with new features and backwards compatibility is maintained, to some degree (long notice periods of breaking changes).

Maybe that’s a change for a later date, when I can work out the best way of implementing it.

Getting rid of the slide shouldn’t be difficult though. I’ll sort it when I get back.

**Comment by boredazfcuk** [comment]
Odd, this is what mine displays:

```
2021-08-27 23:25:51 INFO     Sync user boredazfcuk
2021-08-27 23:25:51 INFO     Synchronisation starting at 23:25:51
2021-08-27 23:25:51 INFO     Check 2FA Cookie
2021-08-27 23:25:51 INFO     Cookie exists, check expiry date
2021-08-27 23:25:51 INFO     Valid two factor authentication cookie found. Days until expiration: 89
2021-08-27 23:25:51 INFO     Check download directory mounted correctly
2021-08-27 23:25:51 INFO     Failsafe file /home/boredazfcuk/iCloud/.mounted exists, continuing.
2021-08-27 23:25:51 INFO     Check for new files using password stored in keyring file.
2021-08-27 23:25:51 INFO     Generating list of files in iCloud. This may take a long time if you have a large photo collection. Please be patient. Nothing is being downloaded at this time
2021-08-27 23:26:08 INFO     Check successful
2021-08-27 23:26:08 INFO     No new files detected. Nothing to download
2021-08-27 23:26:08 INFO     Web cookie expires: 2021-10-26 @ 21:39:28
2021-08-27 23:26:08 INFO     Two factor authentication cookie expires: 2021-11-25 @ 21:39:44
2021-08-27 23:26:08 INFO     Days remaining until expiration: 89
2021-08-27 23:26:08 INFO     iCloud login counter = 1
2021-08-27 23:26:08 INFO     Synchronisation ended at 23:26:08
2021-08-27 23:26:08 INFO     Total time taken: 00:00:17
2021-08-27 23:26:08 INFO     Next synchronisation at 11:25:51
```

This line:
```
echo "$(date '+%Y-%m-%d %H:%M:%S') INFO Total time taken: $(date +%H:%M:%S -d @$((synchronisation_end_time - synchronisation_start_time - 3600)))"
```

Just takes the time the sync ended (in epoch time) and subtracts it from the time that it was started (also epoch). This difference should be the amount of time the sync took in seconds. I then convert that epoch time back to a date... 

Epoch time starts at 1/1/1970 @ 00:00:00. so by converting this epoch time back to a date, I basically only get the amount of seconds, minutes and hours that the sync took... So say a sync took 1384 seconds (23 mins and 4 secs), to complete, converting the number 1384 would give an epoch date of 1 January 1970 00:23:04. I then just format the date so it gives me 00:23:04.

It's a quick and dirty way of converting seconds to hours, minutes and seconds... I'm subtracting an extra hour (-3600) because it was adding one on for some reason. I didn't give it much thought and just presumed I was misremembering epoch start time, as being 1am rather than midnight... I think it's probably the date utility accounting for time zones though. I can't think of any other reason it would be so far out on your machine.

**Comment by bauzer714** [comment]
I think you're missing that epoch is GMT/UTC.. In addition yours is showing 00 hours because you ran the script at 2300. Notice mine is 18xx and the rest of the logs are 17xx?

**Comment by boredazfcuk** [comment]
Yeah, I've pushed a new version to DockerHub which should fix it, I think.

**Comment by bauzer714** [comment]
Cool, appreciate you. I'll pull it over the weekend and let you know.

**Comment by boredazfcuk** [comment]
Basically:

```
Total time taken: $(date +%H:%M:%S -u -d @$((synchronisation_end_time - synchronisation_start_time)))
```

The `-u` should make it output in UTC, ignoring people's timezones... It does with mine anyway.

**Comment by bauzer714** [comment]
That should fix it, thanks again.
